include/boost/lexical_cast/detail/converter_lexical_streams.hpp

```diff
@@ -304,19 +294,19 @@ namespace boost { namespace detail { namespace lcast {
         bool stream_in(lcast::exact<signed char> x)             { return shl_char(static_cast<char>(x.payload)); }
         
         template <class C>
-        typename boost::enable_if_c<boost::detail::is_character<C>::value, bool>::type
+        typename std::enable_if<boost::detail::is_character<C>::value, bool>::type
                 stream_in(lcast::exact<C> x)                    { return shl_char(x.payload); }
 
         template <class Type>
         enable_if_compatible_char_t<Type>
                 stream_in(lcast::exact<Type*> x)                { return shl_char_array(reinterpret_cast<CharT const*>(x.payload)); }
 
         template <class Type>
-        typename boost::enable_if_c<boost::is_signed<Type>::value && !boost::is_enum<Type>::value, bool>::type
+        typename std::enable_if<!std::is_floating_point<Type>::value && std::is_signed<Type>::value && !std::is_enum<Type>::value, bool>::type
```

Comment:
We have to add the `!std::is_floating_point<Type>::value` check here, since `boost::is_signed` and `std::is_signed` behave differently. We have this (which you can double check on Compiler Explorer here: https://godbolt.org/z/xjYWGrTce)

```
static_assert(std::is_signed<float>::value);
static_assert(!boost::is_signed<float>::value);
```

(same for all kind of floating point types).

If we don't change this, then the tests `is_optimized_stream` in `test/stream_traits_test.cpp` are failing for the float to string conversions in optimized mode.